[SPARK-11562][SQL] Provide option to switch SqlContext/HiveContext for spark shell

xguo27 opened the pull request:
Introducing a boolean property 'spark.sql.hive.context' to turn HiveContext on and off as the default sqlContext type.

Commit:
[SPARK-11562][SQL] Provide option to switch SQLContext/HiveContext for Spark shell

AmplabJenkins commented:
Can one of the admins verify this patch?

zhzhan commented on repl/scala-2.10/src/main/scala/org/apache/spark/repl/SparkILoop.scala:
```diff
@@ -132,6 +132,7 @@ class SparkILoop(
   @DeveloperApi
   var sparkContext: SparkContext = _
   var sqlContext: SQLContext = _
+  var sqlContextWithHive: Boolean = _
```
I don't think we need an extra var for this, and we can get the config directly since nowhere else will use it.

zhzhan commented on repl/scala-2.11/src/main/scala/org/apache/spark/repl/Main.scala:
```diff
@@ -78,16 +79,21 @@ object Main extends Logging {
+      logInfo("Created sql context with " + name)
     } catch {
       case _: java.lang.ClassNotFoundException | _: java.lang.NoClassDefFoundError =>
         sqlContext = new SQLContext(sparkContext)
```
I don't know how to fix this, but the logic seems to be weird if SQLContext is used, then in the exception handler the context is created again.

yhuai replied:
I guess we just do not need to catch anything if we use SQLContext?

liancheng replied:
I think it makes sense to just abort if we fail to instantiate the context instance. But should provide informative information and suggestion about 1) the newly added option, and 2) if `HiveContext` is required, ask the user to build Spark with `-Phive`.

zhzhan commented:
Need document update for this new configuration.

xguo27 commented:
Hi Zhan:

I just updated documentation and added a guard in the code regarding your feedback on the exception handler.

Thanks!

xguo27 commented:
@marmbrus @rxin Does this look good to you guys?

yhuai commented on repl/scala-2.10/src/main/scala/org/apache/spark/repl/SparkILoop.scala:
```diff
@@ -1015,21 +1016,27 @@ class SparkILoop(
   @DeveloperApi
   def createSQLContext(): SQLContext = {
-    val name = "org.apache.spark.sql.hive.HiveContext"
+    val name = sqlContextWithHive match {
```
We can get the conf value at here, right?

liancheng commented:
ok to test

liancheng commented on repl/scala-2.10/src/main/scala/org/apache/spark/repl/SparkILoop.scala:
```diff
@@ -1015,21 +1016,27 @@ class SparkILoop(
+    val name = sqlContextWithHive match {
+      case true => "org.apache.spark.sql.hive.HiveContext"
+      case false => "org.apache.spark.sql.SQLContext"
+    }
```
Nit: Why not just an `if` here?

liancheng commented on docs/configuration.md:
```diff
@@ -192,6 +192,13 @@ of the most common options to set are:
   </td>
 </tr>
+<tr>
+  <td><code>spark.sql.hive.context</code></td>
```
How about renaming this option to `spark.sql.useHiveContext`?

SparkQA commented:
**[Test build #46875 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/46875/consoleFull)** for PR 9553 at commit [`888bf4e`](https://github.com/apache/spark/commit/888bf4e3221d6131ebc86fb9af05abccc843c70f).

SparkQA commented:
**[Test build #46875 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/46875/consoleFull)** for PR 9553 at commit [`888bf4e`](https://github.com/apache/spark/commit/888bf4e3221d6131ebc86fb9af05abccc843c70f).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/46875/
Test PASSed.

xguo27 commented:
Hi @yhuai @liancheng:

As I was hitting SPARK-11112 when testing my code, I rebased my branch and squashed my previous commits together. Now the new commit addresses the following point you brought up:

1. call conf.getBoolean() to get the conf value at the right place
2. using spark.sql.useHiveContext instead of spark.sql.hive.context
3. using if/else instead of cases true/false
4. provide extra information when logging

Thanks for reviewing my code!

SparkQA commented:
**[Test build #46905 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/46905/consoleFull)** for PR 9553 at commit [`ce01200`](https://github.com/apache/spark/commit/ce01200604c6feca7cd68918a2f3e15c433827d8).

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/46904/
Test FAILed.

xguo27 commented:
Looks like some git plugin network issue?

yhuai commented:
test this please

SparkQA commented:
**[Test build #46906 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/46906/consoleFull)** for PR 9553 at commit [`ce01200`](https://github.com/apache/spark/commit/ce01200604c6feca7cd68918a2f3e15c433827d8).

SparkQA commented:
**[Test build #46905 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/46905/consoleFull)** for PR 9553 at commit [`ce01200`](https://github.com/apache/spark/commit/ce01200604c6feca7cd68918a2f3e15c433827d8).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:\n  * `case class Count(children: Seq[Expression]) extends DeclarativeAggregate `\n

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/46905/
Test FAILed.

SparkQA commented:
**[Test build #46906 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/46906/consoleFull)** for PR 9553 at commit [`ce01200`](https://github.com/apache/spark/commit/ce01200604c6feca7cd68918a2f3e15c433827d8).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/46906/
Test PASSed.

xguo27 commented:
Hi @yhuai, do you think this is good to merge?

andrewor14 commented:
I'm not sure if I see the use case. If you don't want a hive context just don't build your jar with `-Phive`? If possible it would be best to not introduce an arbitrary config like the one added here.

tgravescs commented:
The use case here is that I want to build one common spark that is used across many different clusters.  Those clusters may not have Hive running yet, or perhaps I just don't want to have to specify hive-site.xml and include hcatalog, etc.  If I'm not using Hive then I don't need spark-shell to load it for me.  Perhaps I'm trying to use Spark and they take Hive down to do maintenance, now I can't run Spark or I get error messages out.

yhuai commented:
@xguo27 can you resolve the conflict?

xguo27 commented:
@yhuai I just resolved the conflict. Can we trigger a test? Thanks!

SparkQA commented:
**[Test build #48049 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/48049/consoleFull)** for PR 9553 at commit [`8395411`](https://github.com/apache/spark/commit/83954118c6151ec2ae83867d44fa482bac30afd4).

SparkQA commented:
**[Test build #48049 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/48049/consoleFull)** for PR 9553 at commit [`8395411`](https://github.com/apache/spark/commit/83954118c6151ec2ae83867d44fa482bac30afd4).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/48049/
Test PASSed.

yhuai commented on repl/scala-2.10/src/main/scala/org/apache/spark/repl/SparkILoop.scala:
```diff
@@ -1026,17 +1027,30 @@ class SparkILoop(
         sqlContext = new SQLContext(sparkContext)
-        logInfo("Created sql context..")
+        logInfo("Created sql context without Hive support, " +
+          "build Spark with -Phive to enable Hive support.")
```
`Created sql context without Hive support. To enable Hive support, build Spark with -Phive profile.`?

yhuai replied:
It is possible that `-Phive` is used but there was an exception. Can we also update the log message for this case?

xguo27 replied:
When -Phive is used (which provides necessary hive jars) and an exception other than ClassNotFound/NoClassDefFound occured, now how we handle it is to let the exception be propagated without creating an alternative SqlContext.

Do you mean by this case, we should catch -> log -> re-throw?

yhuai commented on repl/scala-2.10/src/main/scala/org/apache/spark/repl/SparkILoop.scala:
```diff
@@ -132,6 +132,7 @@ class SparkILoop(
   @DeveloperApi
   var sparkContext: SparkContext = _
   var sqlContext: SQLContext = _
+  var useHiveContext: Boolean = _
```
Can we make it a local field in `createSQLContext`?

xguo27 committed:
SPARK-11562 log message improvement

SparkQA commented:
**[Test build #48057 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/48057/consoleFull)** for PR 9553 at commit [`969008e`](https://github.com/apache/spark/commit/969008ef85723eb35d4ab038c9158eb3ea12b031).

SparkQA commented:
**[Test build #48057 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/48057/consoleFull)** for PR 9553 at commit [`969008e`](https://github.com/apache/spark/commit/969008ef85723eb35d4ab038c9158eb3ea12b031).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/48057/
Test PASSed.

AmplabJenkins commented:
Can one of the admins verify this patch?

andrewor14 commented:
@xguo27 HiveContext is now removed and SQLContext is no longer the default thing. The spark-shell now uses `SparkSession` instead. The changes here are mostly outdated so let's close this PR.

xguo27 commented:
Yeah sure @andrewor14
